kernel/objects/capability-spinning/objects/RevokeOperation.cc

```diff
@@ -99,98 +99,92 @@ namespace mythos {
     _startAsyncDelete(t);
   }
 
   optional<void> RevokeOperation::_delete(CapEntry* root, Cap rootCap)
   {
+    MLOG_ERROR(mlog::cap, __PRETTY_FUNCTION__, DVAR(root), DVAR(rootCap));
     CapEntry* leaf;
+    Cap leafCap;
     do {
-      if (_startTraversal(root, rootCap)) {
-        leaf = _findLockedLeaf(root);
-        MLOG_DETAIL(mlog::cap, "_findLockedLeaf returned", DVAR(*leaf), DVAR(rootCap));
-        if (leaf == root && !rootCap.isZombie()) {
-          // this is a revoke, do not delete the root. no more children -> we are done
-          root->finishRevoke();
-          root->unlock();
-          root->unlock_prev();
-          RETURN(Error::SUCCESS);
-        }
-        auto leafCap = leaf->cap();
-        ASSERT(leafCap.isZombie());
-        if (leafCap.getPtr() == _guarded) {
-          leaf->unlock();
-          leaf->unlock_prev();
-          // attempted to delete guarded object
-          THROW(Error::CYCLIC_DEPENDENCY);
-        }
-        auto delRes = leafCap.getPtr()->deleteCap(*leaf, leafCap, *this);
-        if (delRes) {
-          leaf->unlink();
-          leaf->reset();
-        } else {
-          // Either tried to delete a portal that is currently deleting
-          // or tried to to delete _guarded via a recursive call.
-          leaf->unlock();
-          leaf->unlock_prev();
-          RETHROW(delRes);
-        }
-      } else RETURN(Error::SUCCESS); // could not restart, must be done
+      // compare only value, not the zombie state
+      if (root->cap().asZombie() != rootCap.asZombie()) {
+        // start has a new value
+        // must be the work of another deleter ... success!
+        RETURN(Error::SUCCESS); // could not restart, must be done
+      }
+      if (!_findLeaf(root, rootCap, leaf, leafCap)) continue;
+      MLOG_ERROR(mlog::cap, "_findLockedLeaf returned", DVAR(*leaf), DVAR(rootCap));
+      if (leaf == root && !rootCap.isZombie()) {
+        // this is a revoke, do not delete the root. no more children -> we are done
+        root->finishRevoke();
+        RETURN(Error::SUCCESS);
+      }
+      ASSERT(leafCap.isZombie());
+      if (leafCap.getPtr() == _guarded) {
+        leaf->unlock_cap();
+        // attempted to delete guarded object
+        THROW(Error::CYCLIC_DEPENDENCY);
+      }
+      leaf->lock_cap();
+      if (leaf->cap() != leafCap) {
+        MLOG_DETAIL(mlog::cap, "leaf cap changed concurrently");
+        leaf->unlock_cap();
+        continue;
+      }
+      auto delRes = leafCap.getPtr()->deleteCap(*leaf, leafCap, *this);
+      auto prevres = leaf->lock_prev();
+      ASSERT_MSG(prevres, "somebody unlinked CapEntry currently in unlinking process");
+      leaf->lock_next();
+      if (delRes) {
+        leaf->unlinkAndUnlockLinks();
+        leaf->reset();
+      } else {
+        // deletion failed
+        // Either tried to delete a portal that is currently deleting
+        // or tried to to delete _guarded via a recursive call.
```

Comment:
Sorry, that is the same thing ... the portal is the _guarded`object. I will rephrase that comment.

Reply:
yes, but there is also this REVOKING_FLAG that does something similar to the guarded argument if I understood correctly. I cannot remember what the difference was and why.

Reply:
The `REVOKING_FLAG` is just there to be sure nobody moves a Capability while we try to revoke it (since we (re)start our search for a leaf at the _CapEntry_). Without the _move_ operation, we should not need that flag any more.